modules/api/src/main/java/org/apache/ignite/lang/CancelHandleImpl.java

```diff
@@ -0,0 +1,166 @@
+
+    private final CompletableFuture<Void> cancelFut = new CompletableFuture<>();
+
+    private final CancellationTokenImpl token;
+
+    CancelHandleImpl() {
+        this.token = new CancellationTokenImpl(this);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void cancel() {
+        doCancelAsync();
+
+        cancelFut.join();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CompletableFuture<Void> cancelAsync() {
+        doCancelAsync();
+
+        return cancelFut;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public boolean isCancelled() {
+        return token.isCancelled();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public CancellationToken token() {
+        return token;
+    }
+
+    private void doCancelAsync() {
+        token.cancel().whenComplete((r, t) -> {
+            if (t != null) {
+                cancelFut.completeExceptionally(t);
+            } else {
+                cancelFut.complete(null);
+            }
+        });
+    }
+
+    static final class CancellationTokenImpl implements CancellationToken {
+
+        private final CancelHandleImpl handle;
+
+        private final Object mutex = new Object();
+
+        // Actions that trigger cancellations - action that triggers a cancellation + a future that completes, when a resource is closed.
+        private final ArrayDeque<Cancellation> cancellations = new ArrayDeque<>();
+
+        private volatile CompletableFuture<Void> cancelFut;
+
+        CancellationTokenImpl(CancelHandleImpl handle) {
+            this.handle = handle;
+        }
+
+        boolean isCancelled() {
+            return cancelFut != null;
+        }
+
+        CompletableFuture<Void> cancelHandleFut() {
+            return handle.cancelFut;
+        }
+
+        void addCancelAction(Runnable cancelAction, CompletableFuture<?> completionFut) {
+            assert cancelAction != null : "cancelAction must be not null";
+            assert completionFut != null : "completionFut must be not null";
+
+            Cancellation cancellation = new Cancellation(cancelAction, completionFut);
+
+            synchronized (mutex) {
+                if (cancelFut == null) {
+                    cancellations.add(cancellation);
+                    return;
+                }
+            }
+
+            // Run cancellation action outside of lock
+            cancellation.run();
+        }
+
+        @SuppressWarnings("rawtypes")
+        private CompletableFuture<Void> cancel() {
+            CompletableFuture<Void> f = cancelFut;
+
+            if (f != null) {
+                return f;
+            } else {
+                List<Cancellation> registered;
+
+                synchronized (mutex) {
+                    if (cancelFut != null) {
+                        return cancelFut;
+                    }
+
+                    // First assemble all completion futures
+                    registered = new ArrayList<>(cancellations);
```

Comment:
what is the reason for using a new array instead of iterating through `cancellations`?

Reply:
Fixed.